modules/bullet/space_bullet.cpp

```diff
@@ -890,202 +890,150 @@ void SpaceBullet::update_gravity() {
+		if (is_colliding) {
 
-		has_penetration = recover_from_penetration(p_body, body_transform, 1, p_infinite_inertia, __rec, &r_recover_result);
-
-		// Parse results
-		if (r_result) {
-			B_TO_G(motion + initial_recover_motion + __rec, r_result->motion);
+			btVector3 discard_motion(0, 0, 0);
+			btTransform collision_transform = body_transform;
+			if (motion.fuzzyZero()) {
+				btVector3 original_motion;
+				G_TO_B(p_motion, original_motion);
+				collision_transform.getOrigin() += original_motion * 0.1f;
+			} else {
+				collision_transform.getOrigin() += motion;
+			}
 
-			if (has_penetration) {
+			KinematicCollisionResult collision_result;
+			if (kinematic_collision_test(p_body, collision_transform, p_infinite_inertia, discard_motion, &collision_result)) {
 
-				const btRigidBody *btRigid = static_cast<const btRigidBody *>(r_recover_result.other_collision_object);
+				// Extract collision information.
+				const btRigidBody *btRigid = static_cast<const btRigidBody *>(collision_result.other_collision_object);
 				CollisionObjectBullet *collisionObject = static_cast<CollisionObjectBullet *>(btRigid->getUserPointer());
 
-				B_TO_G(motion, r_result->remainder); // is the remaining movements
-				r_result->remainder = p_motion - r_result->remainder;
-
-				B_TO_G(r_recover_result.pointWorld, r_result->collision_point);
-				B_TO_G(r_recover_result.normal, r_result->collision_normal);
-				B_TO_G(btRigid->getVelocityInLocalPoint(r_recover_result.pointWorld - btRigid->getWorldTransform().getOrigin()), r_result->collider_velocity); // It calculates velocity at point and assign it using special function Bullet_to_Godot
-				r_result->collider = collisionObject->get_self();
-				r_result->collider_id = collisionObject->get_instance_id();
-				r_result->collider_shape = r_recover_result.other_compound_shape_index;
-				r_result->collision_local_shape = r_recover_result.local_shape_most_recovered;
-
-#if debug_test_motion
-				Vector3 sup_line2;
-				B_TO_G(motion, sup_line2);
-				normalLine->clear();
-				normalLine->begin(Mesh::PRIMITIVE_LINES, NULL);
-				normalLine->add_vertex(r_result->collision_point);
-				normalLine->add_vertex(r_result->collision_point + r_result->collision_normal * 10);
-				normalLine->end();
-#endif
-			} else {
-				r_result->remainder = Vector3();
+				B_TO_G(collision_result.pointWorld, r_motion_result->collision_point);
+				B_TO_G(collision_result.normal, r_motion_result->collision_normal);
+				B_TO_G(btRigid->getVelocityInLocalPoint(collision_result.pointWorld - btRigid->getWorldTransform().getOrigin()), r_motion_result->collider_velocity);
+				r_motion_result->collider = collisionObject->get_self();
+				r_motion_result->collider_id = collisionObject->get_instance_id();
+				r_motion_result->collider_shape = collision_result.other_compound_shape_index;
+				r_motion_result->collision_local_shape = collision_result.local_shape_most_recovered;
+
+				// Use binary search to find the safe motion.
```

Comment:
This section seems strange. Why is a binary search used to find the safe motion? Shouldn't btResult.m_closestHitFraction already contain the safe motion? And is that value not already applied to motion from this code block:

```
// Adjust the motion so only closer shapes will also hit.
motion *= btResult.m_closestHitFraction;
```

Reply:
I dont understand why there is a "step 3" that has to call "kinematic_collision_test" again.

Why is the algorithm incomplete after performing `dynamicsWorld->convexSweepTest`?

Reply:
The `btResult.m_closestHitFraction` returns the proportion of the move that will result in a collision, we need to find the proportion of the move that will **not** result in a collision: the safe distance. Furthermore, the `convexSweepTest` doesn't detect all collisions; so sometimes `btResult.m_closestHitFraction` is 1.0. A binary search is an effective means to find this proportion to any level of accuracy. With 8 steps (used in Godot physics too) the precision is 0.39% of the total move distance.

Reply:
> The btResult.m_closestHitFraction returns the proportion of the move that will result in a collision, we need to find the proportion of the move that will not result in a collision

Could this be `btResult.m_closestHitFraction - 0.001` ?

> Furthermore, the convexSweepTest doesn't detect all collisions

Which collisions are not detected by convex sweep test? Concave shapes?

Reply:
> Could this be btResult.m_closestHitFraction - 0.001 ?

This would only work if we knew that `m_closestHitFraction` had that level of accuracy, which we don't; if only, because it sometimes returns 1.0.

> Which collisions are not detected by convex sweep test? Concave shapes?

I haven't done a thorough analysis, but it was more than just concave shapes, which is why I included the binary search.